text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+        identity.attribute_exists(attrName,attrValue);
+        identity.attribute_value(attrName);
+        identity.getID()
+        identity.getMSPID();
+```
+
+Will validate that this is the appropriate set of the accessors to support against real use-cases. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Wasm Runtime
+
+These are the major components within the runtime.
+
+- a Wasm runtime engine responsible for the execution of the Wasm binary
+- grpc communications with the peer
+- data marshalling for transfer of data and requests into and out of the Wasm engine
+- Along with required code to bootstrap the whole Wasm engine and load in the binary
+
+### Wasm Engine
+Several different implementations are available; initial prototypes have used the LifeVM, but for final implementation, Wasmer is being considered. The WebAssembly landscape is still developing so there is some risk with changing standards and implementations, however, some of the new standards could be of benefit in the future.
+
+### grpc Peer Communications
+This will use the go chaincode shim today to provide connectivity to the peer.
+
+### Data Marshalling
+Currently, the API to import/export functions and data between the host and the executing code in the wasm runtime is limited to primitive data types only. A data and function marshalling interface is required, waPC will be used to provide this. 
+
+The logical functions that will be marshalled over waPC from host to client will form part of the deliverables. This is to allow for new languages to be supported and also to permit changes to the peer communications protocol. (these logical functions are referred to as the Contract Core SPI). It will be designed to use the protobuf specification to permit accurate description.
+
+## Lifecycle of the Wasm binary
+
+The Wasm binary that is produced will be an input to the normal chaincode lifecycle with supplied Wasm external chaincode builder scripts. 
+
+The Wasm binary will not be stored within the world state; on-chain storage and execution of WebAssembly smart contracts are beyond the scope of this RFC.
+
+## Contract and Ledger APIs
+
+A prototype implementation of the interfaces is available in [github](https://github.com/hyperledgendary/fabric-contract-api-rust) 
+
+The core functionality of Fabric, previously available in the shim and stub APIs, is replicated within this RFC. 
+
+# Repositories and Deliverables
+
+It is anticipated that the following github repositories will be created.
+
+- `fabric-builder-wasm`
+    - external builder scripts and Dockerfile for a Wasm aware peer
+- `fabric-chaincode-wasm`
+    - the chaincode as a server to host wasm contracts
+- `fabric-contract-api-rust`  
+    - contracts 
+- `fabric-ledger-protos`      
+    - source for protos
+- `fabric-ledger-protos-go`   
+    - auto-generated from protos repo
+- `fabric-ledger-protos-rust` 
```

Comment:
We already have too many repos. Not sure whether 6 more repos are needed.

Reply:
possibly bad phrasing - consider these the main components of the system; they may in their repos or not.

Reply:
I see. Got it. Thanks, Matthew.